sw/device/tests/rstmgr_cpu_info_test.c

```diff
@@ -0,0 +1,202 @@
+#include "sw/device/lib/testing/test_framework/ottf_macros.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+/**
+ * RSTMGR CPU INFO TEST
+ *  This test create double fault by accessing register with
+ *  non-existing address.
+ *  After double fault, dut got reset by watch dog bite
+ *  and test collect / check cpu_info from rstmgr
+ */
+
+// non existing address
+#define kIllegalAddr1 0x4041FFF0u
+#define kIllegalAddr2 0x40003618u
+#define kSkipComp 0x12345678u
+#define kCpuDumpSize 9
+
+/**
+ * Cpu dump struct index
+ */
+enum {
+  kCpuDumpIdxCurrentExceptionAddr = 0,
+  kCpuDumpIdxCurrentLastDataAddr = 1,
+  kCpuDumpIdxCurrentNextPc = 2,
+  kCpuDumpIdxCurrentPc = 3,
+  kCpuDumpIdxPreviousExceptionAddr = 4,
+  kCpuDumpIdxPreviousLastDataAddr = 5,
+  kCpuDumpIdxPreviousNextPc = 6,
+  kCpuDumpIdxPreviousPc = 7,
+  kCpuDumpIdxPreviousValid = 8,
+};
+
+/**
+ * Reserve expected cpu dump area in flash
+ */
+__attribute__((section(".non_volatile_scratch")))
+const volatile dif_rstmgr_cpu_info_dump_segment_t exp_dump[kCpuDumpSize] = {
+    UINT32_MAX, UINT32_MAX, UINT32_MAX, UINT32_MAX, UINT32_MAX,
+    UINT32_MAX, UINT32_MAX, UINT32_MAX, UINT32_MAX};
+
+/**
+ *  Dump structure:
+ *    0: current.exception_addr
+ *    1: current.last_data_addr
+ *   *2: current.next_pc
+ *   *3: current.pc
+ *    4: previous.exception_addr
+ *    5: previous.last_data_addr
+ *   *6: previous.next_pc
+ *    7: previous.pc
+ *    8: previous_valid
+ *
+ * Observed cpu dump will be collected after watch dog bite,
+ * exp cpu dump will be created at the 'ottf_exception_handler'.
+ * Following fields are current code specific and
+ * will be skipped comparison.
+ *
+ *  current.next_pc, current.pc, previous.next_pc
+ */
+static dif_rstmgr_cpu_info_dump_segment_t dump[kCpuDumpSize];
+static dif_rstmgr_cpu_info_dump_segment_t temp_dump[kCpuDumpSize] = {
+    kSkipComp, kSkipComp, kSkipComp, kSkipComp, kSkipComp,
+    kSkipComp, kSkipComp, kSkipComp, kSkipComp};
+
+static dif_flash_ctrl_state_t flash_ctrl;
+
+// Count number of falut
+static uint32_t global_error_cnt;
+
+// Access non-existing address
+// Each call will create a fault
+static void read_error(void) {
+  uint32_t addr;
+  global_error_cnt++;
+
+  if (global_error_cnt == 1) {
+    addr = kIllegalAddr1;
+  } else {
+    LOG_INFO("double fault");
+    addr = kIllegalAddr2;
+  }
+  // I can't add a new variable to call mmio_
+  // cause mmio call will never be returned.
+  // Use current variable, just to avoid unused error.
+  addr = mmio_region_read32(mmio_region_from_addr(addr), 0);
+}
+
+/**
+ * Overrides the default OTTF exception handler.
+ *
+ */
+void ottf_exception_handler(void) {
+  uint32_t addr = (uint32_t)exp_dump;
+
+  temp_dump[kCpuDumpIdxCurrentExceptionAddr] =
+      (dif_rstmgr_cpu_info_dump_segment_t)ibex_mepc_read();
+  temp_dump[kCpuDumpIdxPreviousExceptionAddr] =
+      temp_dump[kCpuDumpIdxCurrentExceptionAddr];
```

Comment:
am I missing something here or should these lines come before lines 112-113? I.e. would i want to update the previous before the current? otherwise both the previous and current will always be equal.

Reply:
In this test, they are always equal.